optimum/onnxruntime/modeling_ort.py

```diff
@@ -200,11 +200,15 @@ def __init__(
                 f" Use `ort_model.to()` to send the outputs to the wanted device."
             )
 
-        if "TensorrtExecutionProvider" in self.providers and self.use_io_binding:
-            logger.warning(
-                "There is no need to do IO binding for TensorrtExecutionProvider, `use_io_binding` is set to False."
-            )
-            self.use_io_binding = False
+        io_binding_options = possible_io_binding_for_provider(self.providers)
```

Comment:
why not

```python
if provider == "CUDAExecutionProvider" and use_io_binding is None:
    use_io_binding = True
elif provider != "CUDAExecutionProvider" and use_io_binding is True:
    logger.warning("Asked to use IO Binding although the provider used is not CUDAExecutionProvider. IO Binding will not be used.")
    use_io_binding = False
else:
    use_io_binding = False
```

Reply:
I opened this PR as I don't want to decide the value of `use_io_binding` by `provider`. E.g. one can try loading the inference session with TRT and it falls back to CUDA ep. It would be better to set the io binding depends on what ep are truly used in the end.

Reply:
> E.g. one can try loading the inference session with TRT and it falls back to CUDA ep.

This should not happen if it's an issue with the TRT install, as per https://github.com/huggingface/optimum/pull/514

The possible fallback is only in some nodes in the ONNX.

Reply:
Whether the ep falls back to another ep or not, the error message should not be about io binding. Set the value according to the truly used eps avoid the misleading error message.

Reply:
there's no such thing as "truly used eps". ONNX Runtime can mix execution providers depending on which EPs support which nodes.